File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def dfs(numbers, target, depth):
    answer = 0
    
    if depth == len(numbers):
        if sum(numbers) == target:
            return 1
        return 0
    
    answer += dfs(numbers, target, depth + 1)
    numbers[depth] *= -1
    answer += dfs(numbers, target, depth + 1)
    
    return answer
    
    
def solution(numbers, target):
    return dfs(numbers, target, 0)
    
    

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    public static int count = 0;
    
    public void dfs(int sum, int[] numbers, int idx, int target) {
        if (idx == numbers.length) {
            if (sum == target) count += 1;
            return;
        }
        
        dfs(sum + numbers[idx], numbers, idx + 1, target);
        dfs(sum - numbers[idx], numbers, idx + 1, target);
    }
    
    public int solution(int[] numbers, int target) {
        dfs(0, numbers, 0, target);
        return count;
    }
}